src/main/java/org/jitsi/videobridge/RtpChannel.java

```diff
@@ -1991,6 +1968,121 @@ public RtpChannelTransformEngine getTransformEngine()
         return this.transformEngine;
     }
 
+    /**
+     * Creates the {@code MediaStreamTrack}s from signaling and adds them to the
+     * {@code MediaStream} that is associated to this {@code RtpChannel}.
+     *
+     * @param sources  The <tt>List</tt> of <tt>SourcePacketExtension</tt> that
+     * describes the list of sources of this <tt>RtpChannel</tt> and that is
+     * used as the input in the update of the Sets the <tt>Set</tt> of the SSRCs
+     * that this <tt>RtpChannel</tt> has signaled.
+     * @param sourceGroups
+     */
+    public void setMediaStreamTracks(
+        List<SourcePacketExtension> sources,
+        List<SourceGroupPacketExtension> sourceGroups)
+    {
+        if ((sources == null || sources.isEmpty())
+            && (sourceGroups == null || sourceGroups.isEmpty()))
+        {
+            return;
+        }
+
+        this.setSources(sources); // TODO remove and rely on MSTs.
+        this.setSourceGroups(sourceGroups); // TODO remove and rely on MSTs.
+
+        Map<Long, MediaStreamTrack> tracksBySSRC = new TreeMap<>();
+        Map<Long, long[]> encodings = new TreeMap<>();
+
+        List<Long> freeSSRCs = new ArrayList<>();
+
+        if (sources != null && sources.size() != 0)
+        {
+            for (SourcePacketExtension spe : sources)
+            {
+                freeSSRCs.add(spe.getSSRC());
+            }
+        }
+
+        if (sourceGroups != null && sourceGroups.size() != 0)
+        {
+            for (SourceGroupPacketExtension sgpe : sourceGroups)
+            {
+                if ("sim".equalsIgnoreCase(sgpe.getSemantics())
+                    && sgpe.getSources() != null
+                    && sgpe.getSources().size() >= 2)
+                {
+                    // Every simulcast group determines an mst with a bunch of
+                    // encodings.
+                    MediaStreamTrack track = new MediaStreamTrack();
+
+                    int order = RTPEncoding.BASE_ORDER;
+                    for (SourcePacketExtension spe : sgpe.getSources())
+                    {
+                        long primarySSRC = spe.getSSRC();
+                        freeSSRCs.remove(primarySSRC);
+
+                        long[] encoding = new long[] { -1, -1, -1 };
+                        encoding[0] = primarySSRC;
+                        encoding[1] = order++;
```

Comment:
Ugh, doesn't Java allow you to do the initialization without constants? For example,

```` java
long[] encoding = new long[] { primarySSRC, order++, -1 };
````

I mean, there are magic indices, `-1` that get immediately overwritten.... Even the following would be better

```` java
long[] encoding = new long[3];
encoding[0] = primarySSRC;
encoding[1] = order++;
encoding[2] = -1;
````